Add Core SQL Provider

denimalpaca opened the pull request:
Adds operators, tests, and new and updated docs for a Core SQL
Provider. This provider is made in favor of adding these operators
to the existing SQL operators in core Airflow. The new provider
will allow for quicker development cycles.

closes: #23874, #24422
related: #23915

Commit:
Add Core SQL Provider

Adds operators, tests, and new and updated docs for a Core SQL
Provider. This provider is made in favor of adding these operators
to the existing SQL operators in core Airflow. The new provider
will allow for quicker development cycles.

Commit:
Re-remove core.sql from list of provider tags

I am bad at git and this change got overwritten during a git pull
with the remote branch.

Add Core SQL Provider

Adds operators, tests, and new and updated docs for a Core SQL
Provider. This provider is made in favor of adding these operators
to the existing SQL operators in core Airflow. The new provider
will allow for quicker development cycles.

Fix build doc errors

Fix capitalization and grammar errors in doc

Commit:
Fix build doc errors

Commit:
Fix example DAG provider import statement

Commit:
Add parse_boolean() to operator file to resolve import error

Commit:
Add leq and geq to spelling_wordlist

Commit:
Fix capitalization and grammar errors in doc

Commit:
Remove core.sql from list of tags, use software tag instead for provider

Commit:
Fix broken link reference in provider doc

Commit:
Update breeze images

Commit:
Remove merge conflict lines fix after rebase

ephraimbuddy reviewed (commented):
I don't think the location `/providers/core` is appropriate because `core` doesn't really give much information. Maybe we should move all the SQL operators in `airflow.operator` into the providers package at `/providers/sql`? 
Though, I don't really know why the SQL operators are still in core airflow, so it's something to check

ephraimbuddy commented on airflow/providers/core/sql/operators/sql.py:
```diff
@@ -0,0 +1,308 @@
+                    <= record
+                    <= check_values["equal_to"] * (1 + tolerance)
+                )
+        return record == check_values["equal_to"]
```
```suggestion
            return record == check_values["equal_to"]
        return False
```

denimalpaca replied:
So, this is actually causing a static check error as the function needs a non-nested return statement; as I put in my commit message to make this change, it doesn't actually matter where the return statement is logic-wise because there is already a check that ensures that `check_values` contains some correct value, so having a catch-all return value at the end with `equal_to` should be fine.

ephraimbuddy replied:
This will likely cause error. You can instead return False  at the end but nest this return statement.

ephraimbuddy commented on airflow/providers/core/sql/operators/sql.py:
```diff
@@ -0,0 +1,308 @@
+            and "less_than" not in check_values
+            and "leq_than" not in check_values
+            and "equal_to" not in check_values
+        ):
```
Should `geq_than` and `leq_than` be renamed to `geq_to` and `leq_to` respectively? I'm assuming that this `geq_than` means `greater_than_or_equal_to`?

denimalpaca replied:
I think that's a good change; grammatically it does make more sense.

ephraimbuddy reviewed (commented):
@potiuk, your explanation makes sense to me and I can't think of a better name.

Commit:
Change names of geq_than and leq_than to geq_to and leq_to, respectively

Commit:
Update airflow/providers/core/sql/operators/sql.py

This shouldn't actually matter; because the `check_values` options are checked before this code is executed, there will always be some path to exit the function, whether it's a catch-all `return` at the end of the function as it was, or whether `equal_to` is checked last and the inner `if` doesn't run like the change suggests.

Co-authored-by: Ephraim Anierobi <splendidzigy24@gmail.com>

Commit:
Unindent return in _get_match()

Static checks require an always-accessible return statement.

Commit:
Return False at end of _get_match() to make static checks pass

Commit:
Update breeze output commands hash

Commit:
Fix _get_match() logic

The logic in _get_match() was previously short-circuiting after one
value check, however, multiple value checks per check type are
permitted. The logic is updated to ensure all value checks are
run and that any failure is propogated through the subsequent
checks. Additional tests are added to ensure proper logic.

Commit:
Fix output-commands-hash file after rebase

kaxil commented on airflow/providers/core/__init__.py:
```diff
@@ -0,0 +1,16 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
I am not a fan of "core.sql" too like @ephraimbuddy -- maybe "base.sql" is better 🤷 but I do know that there already is a BaseSQLOperator -- which we should think of moving to a provider too tbh.

But naming is heard. tagging a few people if they have better suggestions for names: @dstandish @jedcunningham @uranusjr

jedcunningham replied:
Maybe "common" would be better than "core"?

potiuk replied:
We can still change it - before we release the provider and I think we should only do it after we move DBApiHook there (as explained in #23971). 

I think common might be a better choice indeed.

kaxil commented on airflow/providers/core/sql/operators/sql.py:
```diff
@@ -0,0 +1,315 @@
+            if not records:
+                raise AirflowException(f"The following query returned zero rows: {self.sql}")
+
+            self.log.info(f"Record: {records}")
```
```suggestion
            self.log.info("Record: %s", records)
```

kaxil commented on airflow/providers/core/sql/operators/sql.py:
```diff
@@ -0,0 +1,315 @@
+    :param table: the table to run checks on
+    :param checks: the dictionary of checks, e.g.:
+
+    .. code-block:: python
+
+        {
+            "row_count_check": {"check_statement": "COUNT(*) = 1000"},
+            "column_sum_check": {"check_statement": "col_a + col_b < col_c"},
+        }
+
+    :param conn_id: the connection ID used to connect to the database
```
I think this should be indented further, not sure though, can you post a screenshot of how this looks in docs

denimalpaca replied:
Is that built if I run `breeze build-docs`? I can't seem to find a preview anywhere (and not sure that putting in a separate markdown file to preview is going to be an accurate reflection of this).

potiuk replied:
Yes. when you run `breeze build-docs --package-filter apache-airlfow-provider-core-sql` and run `./docs/start-docs-server.sh` or something - you should see the docs (it's described in contributing docs I am quite sure of that)

kaxil commented on airflow/providers/core/sql/operators/sql.py:
```diff
@@ -0,0 +1,315 @@
+        if not records:
+            raise AirflowException(f"The following query returned zero rows: {self.sql}")
+
+        self.log.info(f"Record: {records}")
```
```suggestion
        self.log.info("Record: %s", records)
```

kaxil reviewed (commented):
Some minor comments, going to merge this, feel free to take care of the other comments in a follow-up PR please
